internal/erasurecoding/reedsolomon/reedsolomon.go

```diff
@@ -0,0 +1,247 @@
+func (r *Encoder) Decode(shards [][]byte) error {
+	if shardCount(shards) < r.originalShardsCount {
+		return errors.New("too few shards")
+	}
+	shardSize := shardSize(shards)
+	if shardSize == 0 {
+		// todo better error name
+		return errors.New("invalid shard size")
+	}
+
+	flatOriginalShards := []byte{}
+	flatOriginalShardsIndexes := []C.size_t{}
+	for i, s := range shards[:r.originalShardsCount] {
+		if len(s) != 0 {
+			flatOriginalShards = append(flatOriginalShards, s...)
+			flatOriginalShardsIndexes = append(flatOriginalShardsIndexes, C.size_t(i))
+		}
+	}
+
+	flatRecoveryShards := []byte{}
+	flatRecoveryShardsIndexes := []C.size_t{}
+	for i, s := range shards[r.originalShardsCount:] {
+		if len(s) != 0 {
+			flatRecoveryShards = append(flatRecoveryShards, s...)
+			flatRecoveryShardsIndexes = append(flatRecoveryShardsIndexes, C.size_t(i))
+		}
+	}
+
+	shardCountOriginal := shardCount(shards[:r.originalShardsCount])
+	// Shards we already have aren't restored.
+	restoredShardsCount := r.originalShardsCount - shardCountOriginal
+	restoredShards := make([]byte, restoredShardsCount*shardSize)
+	restoredShardsIndexes := make([]C.size_t, restoredShardsCount)
+
+	result := reedSolomonDecode(
+		C.size_t(r.originalShardsCount),
+		C.size_t(r.recoveryShardsCount),
+		C.size_t(shardSize),
+		flatOriginalShards,
+		C.size_t(len(flatOriginalShards)),
+		flatOriginalShardsIndexes,
+		flatRecoveryShards,
+		C.size_t(len(flatRecoveryShards)),
+		flatRecoveryShardsIndexes,
+		restoredShards,
+		restoredShardsIndexes)
+	if result != 0 {
+		return errors.New("unable to decode data")
+	}
+
+	for i := 0; i < restoredShardsCount; i++ {
+		start := i * shardSize
+		end := start + shardSize
+		index := int(restoredShardsIndexes[i])
+		shards[index] = restoredShards[start:end]
+	}
+
+	return nil
+}
+
+// Returns the first non nil or length zero shard length.
+func shardSize(shards [][]byte) int {
+	for _, shard := range shards {
+		if len(shard) != 0 {
+			return len(shard)
+		}
+	}
+	return 0
+}
+
+// Returns the count of non nil or length zero shards.
+func shardCount(shards [][]byte) int {
+	count := 0
+	for _, shard := range shards {
+		if len(shard) != 0 {
+			count++
+		}
+	}
+	return count
+}
+
+func init() {
+	// Load the Rust shared library in the init function
+	libPath, err := getErasurecodingLibaryPath()
+	if err != nil {
+		fmt.Println("Failed to load erasure coding library path:", err)
+		os.Exit(1)
+	}
+
+	// Load the Rust shared library
+	lib, err := purego.Dlopen(libPath, purego.RTLD_NOW|purego.RTLD_GLOBAL)
+	if err != nil {
+		fmt.Println("Failed to load erasure coding library:", err)
+		os.Exit(1)
+	}
+
+	// Register the Rust FFI functions with Go using purego
+	purego.RegisterLibFunc(&reedSolomonEncode, lib, "reed_solomon_encode")
+	purego.RegisterLibFunc(&reedSolomonDecode, lib, "reed_solomon_decode")
+}
+
+func getErasurecodingLibaryPath() (string, error) {
```

Comment:
This function is quite similar to `getBandersnatchLibraryPath`. I suggest to move the method to a central/common location and passing the library name as a parameter.

Reply:
Good point. The only problem is the paths are a little different, so you'd also need to pass the path to the lib. I also think we'll probably need to have some sort of config in place later , or some more sophisticated way of finding the library by searching for a few common paths. I think this might be best to defer until then.